File: scripts/boda/lib/icons.lua
-- Icons drawn with ASS shapes, so they never depend on an emoji font.
local M = {}

local function rect(x, y, w, h)
    return string.format("m %.1f %.1f l %.1f %.1f l %.1f %.1f l %.1f %.1f",
        x, y, x + w, y, x + w, y + h, x, y + h)
end

M.rect = rect

function M.play(s)
    return string.format("m 0 0 l 0 %.1f l %.1f %.1f", s, s * 0.88, s / 2)
end

function M.pause(s)
    return rect(s * 0.08, 0, s * 0.28, s) .. " " .. rect(s * 0.62, 0, s * 0.28, s)
end

function M.prev(s)
    return rect(0, 0, s * 0.16, s) .. " " ..
        string.format("m %.1f %.1f l %.1f 0 l %.1f %.1f", s * 0.24, s / 2, s, s, s)
end

function M.next(s)
    return rect(s * 0.84, 0, s * 0.16, s) .. " " ..
        string.format("m 0 0 l %.1f %.1f l 0 %.1f", s * 0.76, s / 2, s)
end

function M.fullscreen(s)
    local t = math.max(2, s * 0.15)
    local a = s * 0.4
    return table.concat({
        rect(0, 0, a, t), rect(0, 0, t, a),
        rect(s - a, 0, a, t), rect(s - t, 0, t, a),
        rect(0, s - t, a, t), rect(0, s - a, t, a),
        rect(s - a, s - t, a, t), rect(s - t, s - a, t, a),
    }, " ")
end

function M.list(s)
    local t = math.max(2, s * 0.16)
    return table.concat({
        rect(0, 0, s, t), rect(0, (s - t) / 2, s, t), rect(0, s - t, s, t),
    }, " ")
end

function M.speaker(s, muted)
    local body = string.format("m 0 %.1f l %.1f %.1f l %.1f 0 l %.1f %.1f l %.1f %.1f l 0 %.1f",
        s * 0.33, s * 0.28, s * 0.33, s * 0.6, s * 0.6, s, s * 0.28, s * 0.67, s * 0.67)
    if muted then
        local t = s * 0.1
        local d = string.format("m %.1f %.1f l %.1f %.1f l %.1f %.1f l %.1f %.1f",
            s * 0.72, s * 0.28, s * 0.72 + t, s * 0.28, s * 1.02, s * 0.72, s * 1.02 - t, s * 0.72)
        local d2 = string.format("m %.1f %.1f l %.1f %.1f l %.1f %.1f l %.1f %.1f",
            s * 1.02 - t, s * 0.28, s * 1.02, s * 0.28, s * 0.72 + t, s * 0.72, s * 0.72, s * 0.72)
        return body .. " " .. d .. " " .. d2
    end
    return body .. " " .. rect(s * 0.72, s * 0.3, s * 0.08, s * 0.4) ..
        " " .. rect(s * 0.88, s * 0.15, s * 0.08, s * 0.7)
end

function M.subtitle(s)
    local t = math.max(1.5, s * 0.1)
    local h = s * 0.74
    return table.concat({
        rect(0, 0, s, t), rect(0, h - t, s, t), rect(0, 0, t, h), rect(s - t, 0, t, h),
        rect(s * 0.18, h * 0.55, s * 0.26, t), rect(s * 0.54, h * 0.55, s * 0.28, t),
    }, " ")
end

return M

File: tools/icons.py
# -*- coding: utf-8 -*-
"""Generate scripts/boda/lib/icons.lua from the Material Symbols SVGs in tools/mdi/.

    python tools/icons.py

To add an icon: drop its SVG in tools/mdi/ (Material Symbols, rounded, filled,
24px, from https://fonts.google.com/icons) and add a line to ICONS below.
"""
import io, os, sys
HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
from svg2ass import convert

# boda name -> Material Symbols file
ICONS = [
    ("play", "play_arrow"), ("pause", "pause"),
    ("prev", "skip_previous"), ("next", "skip_next"),
    ("volume", "volume_up"), ("mute", "volume_off"),
    ("fullscreen", "fullscreen"), ("subtitle", "subtitles"),
    ("list", "format_list_bulleted"), ("color", "palette"),
    ("video", "movie"), ("audio", "graphic_eq"),
    ("chapter", "bookmark"), ("clip", "content_cut"),
    ("save", "save"), ("folder", "folder_open"),
    ("intro", "login"), ("outro", "logout"),
    ("record", "radio_button_checked"), ("loop", "repeat"),
    ("close", "close"), ("clear", "delete_sweep"),
    ("add", "add"), ("mark_a", "first_page"), ("mark_b", "last_page"),
    ("search", "search"), ("settings", "tune"),
]

head = '''-- Icons, drawn as ASS shapes so nothing depends on a font being installed.
--
-- The paths come from Google's Material Symbols (Apache-2.0, rounded, filled),
-- converted from SVG. They are written for a 24x24 box and scaled on use.
-- To change or add one: drop the .svg in tools/mdi/ and run tools/icons.py.

local M = {}

local function rect(x, y, w, h)
    return string.format("m %.1f %.1f l %.1f %.1f l %.1f %.1f l %.1f %.1f",
        x, y, x + w, y, x + w, y + h, x, y + h)
end

M.rect = rect

local PATHS = {
'''

tail = '''}

-- Scaling means rewriting every number in the path, so keep the results: the
-- same handful of icons is redrawn at the same few sizes over and over.
local cache = {}

local function scaled(name, size)
    local d = PATHS[name]
    if not d then return "" end
    size = math.floor((tonumber(size) or 16) * 4 + 0.5) / 4
    local key = name .. ":" .. size
    local hit = cache[key]
    if hit then return hit end
    local f = size / 24
    local out = d:gsub("%-?%d+%.?%d*", function(n)
        return string.format("%.2f", tonumber(n) * f)
    end)
    cache[key] = out
    return out
end

-- icons.play(16) and icons.get("play", 16) both work
function M.get(name, size) return scaled(name, size) end

for name in pairs(PATHS) do
    M[name] = function(size) return scaled(name, size) end
end

-- the one icon that depends on state
function M.speaker(size, muted)
    return scaled(muted and "mute" or "volume", size)
end

return M
'''

lines = []
for name, src in ICONS:
    path = os.path.join(HERE, "mdi", src + ".svg")
    d = convert(path)
    lines.append('    %s = "%s",' % (name, d))

out = head + "\n".join(lines) + "\n" + tail
target = sys.argv[1] if len(sys.argv) > 1 else "icons.lua"
io.open(target, "w", encoding="utf-8", newline="\n").write(out)
print("wrote %s (%d icons, %d bytes)" % (target, len(ICONS), len(out)))
